### PR TITLE
Fix parental encryption status with EQUALS constraint

### PR DESCRIPTION
**Pull request by wwaughjc**
Fixes https://github.com/osquery/osquery/issues/7314

When an EQUALS constraint is applied to the disk_encryption virtual table,
the parent device also needs to be queried. The EQUALS constraint is applied
by the user by including `WHERE name=` in the query. It's also applied
implicitly by JOINing on the table.

### CODE REVIEW

**Review by directionless — COMMENTED** [review]
Hi! I really like this as a continuation of #7209 (I think I dropped the ball in plumbing this) but I wanted to go over what this was doing. I'm not totally sure I understand. And I'm not sure if the approach is right.

If I understand the underlying problem, is that some devices are not directly encrypted, their parent is. So when osquery grabs the row from `block_devices`, it's not going to get the real device, and won't populate encryption correctly.

For context, when I made #7209, I was intending to allow this table to be queried with whatever name you wanted. name, parentname, a device not in the `block_devices` at all... Though reading your PR, I see I fell short. Even with the constraint, it still routes through `block_devices`. So this leads to a question...

Does it make sense to simply call `genFDEStatusForBlockDevice` for rows that are not in `block_devices` ?

If I read this PR correctly, it's taking a very different approach. I think it's:
1. calling `"block_devices", "name", EQUALS, name`
2. If there's a parent, it then calls `"block_devices", "name", EQUALS, parent_name`
3. and eventually calls `genFDEStatusForBlockDevice` for that. 

But, I feel like I have questions about this approach:
1. If I understand correctly, it's like a single pass on what should probably be a recursive loop?
2. Wouldn't we still need to do this parent resolution even if `runSelectAll` happens?
3. Are the resultant uuid values the correct ones? (I'm super confused by that TBH)

**Comment by wwaughjc** [comment]
Thanks for taking a look! I reworked this to recurse, per your suggestion. I don't have enough context to know what the behavior of the table should be in the larger sense, but this PR addresses an internal inconsistency, so hopefully that's a step in the right direction.

File: osquery/tables/system/linux/disk_encryption.cpp [hunk 1]
```diff
@@ -117,23 +119,41 @@ QueryData genFDEStatus(QueryContext& context) {
   std::map<std::string, Row> encrypted_rows;
 
   bool runSelectAll(true);
-  QueryData block_devices;
+  std::list<Row> block_devices;
```
[the remark this answers is not in the document]
Comment (wwaughjc) [reply osquery/tables/system/linux/disk_encryption.cpp]: Ack. This was to be able to `push_front`.

**Comment by wwaughjc** [comment]
Hey @directionless, quick ping. I'm still working through the CLA within my org, but I'm confident it'll get signed eventually. Meanwhile, I'm happy to work through any other feedback you have.

**Comment by mike-myers-tob** [comment]
Hi @wwaughjc -- was there eventually an update about allowing this CLA?

File: osquery/tables/system/linux/disk_encryption.cpp [hunk 2]
```diff
@@ -30,6 +32,33 @@ const std::string kEncryptionStatusNotEncrypted = "not encrypted";
 namespace osquery {
 namespace tables {
 
+// Helper class to track of device parents and ensure that
```
Comment (lhmzhou) [inline osquery/tables/system/linux/disk_encryption.cpp:35]: Helper class to `keep` track of...

**Comment by zwass** [comment]
@wwaughjc I'd love to see this get merged... Any chance you can complete the CLA?

**Comment by buzzdeee** [comment]
this is affecting me as well, so would also love to have this fixed

**Comment by wwaughjc** [comment]
I was able to get the CLA approved, so please take another look!

**Comment by directionless** [comment]
@artemist can I tag you to take a peek at this and see if it help?

**Comment by wwaughjc** [comment]
Friendly bump

**Comment by mbainter** [comment]
I hate to pile on, or risk this being a "+1" type situation, but it has been over three months since the last friendly bump.

I would love to see this merged and released so I can actually monitor for compliance on this.

**Comment by artemist** [comment]
Oh, sorry, I have a separate github account for work so these notifications just get lumped in with junk in my personal inbox. @artemist-work is the one for fleet work and it's fallen through the cracks after some changes to our internal issue tracker.

**Comment by sbrownjc** [comment]
@artemist-work I certainly appreciate things falling through the cracks due to issue tracker changes. Do you have an idea of when this might get looked at? We've run into an issue with another Ubuntu system that this would likely address as well, so we would love to see it released in the wild ASAP (with proper review, of course).

**Comment by Micah-Kolide** [comment]
I pulled this down for some testing and I was able to get the expected results when using a where name equals clause, but not when we land in the `runSelectAll` portion. Running a query where I didn't select by name would max out my test device's cpu and 4GB of memory until the process dies.

I'm not very well versed in c++, but I tried figuring it out myself by making some changes around a couple of areas without success . I might be on the wrong path, but I do believe the issue is in `sortDevicesByDepth`, since that is really the only new difference when hitting the select all from block_devices.

**Comment by rr-robert** [comment]
+1 for this bugfix to be merged @artemist-work I'm not well versed with this project, what needs to happen to merge?

**Comment by directionless** [comment]
Closed in favor of https://github.com/osquery/osquery/pull/8052

**Pull request closed without merging**